File: app/src/main/java/com/example/babajidemustapha/survey/BarChartFragment.java
package com.example.babajidemustapha.survey;


import android.*;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Environment;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.MarkerView;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.formatter.IAxisValueFormatter;
import com.github.mikephil.charting.formatter.IValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.utils.MPPointF;
import com.github.mikephil.charting.utils.ViewPortHandler;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

/**
 * A simple {@link Fragment} subclass.
 */
public class BarChartFragment extends Fragment {

    int survey_id;
    SurveyDatabase db;
    RecyclerView recyclerView;
    List<Question> questions;
    Map<Question,Map<String,Integer>> xyCoord;
    private final int REQUEST_PERMISSIONS_CODE_WRITE_STORAGE = 18880;
    BarChart barChart1;
    public BarChartFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_bar_chart,container,false);
        barChart1 = null;
        survey_id = getActivity().getIntent().getExtras().getInt("ID");
        db = new SurveyDatabase(getContext());
        questions = new ArrayList<>();
        xyCoord = new LinkedHashMap<>();
        Map<Question,List<Response>> reports = db.getReport(survey_id);
        buildXYcoord(reports);
        recyclerView = (RecyclerView) view.findViewById(R.id.barList);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new CustomAdapter1(questions));
        return view;
    }
    public void buildXYcoord(Map<Question,List<Response>> report){

        for (Map.Entry<Question,List<Response>> x : report.entrySet()) {
            Log.e("question NO: ", x.getKey().getQuestionNo()+"");
            switch (x.getKey().getQuestionType()){
                case "TEXT":
                    xyCoord.put(x.getKey(),null);
                    questions.add(x.getKey());
                    break;
                case "SINGLE":
                    Map<String,Integer> xy = new LinkedHashMap<>();
                    for(int i = 0;i<x.getKey().getOptions().size();i++){
                        xy.put(x.getKey().getOptions().get(i),0);
                    }
                    for (Response response: x.getValue()) {
                        if (!response.getResponse().trim().isEmpty()) {
                            if (xy.containsKey(response.getResponse())) {
                                int i = xy.get(response.getResponse()) + 1;
                                xy.put(response.getResponse(), i);
                            } else {
                                xy.put(response.getResponse(), 1);
                            }
                        }
                    }
                    xyCoord.put(x.getKey(),xy);
                    questions.add(x.getKey());
                    break;
                case "MULTI":
                    Map<String,Integer> xy1 = new LinkedHashMap<>();
                    for(int i = 0;i<x.getKey().getOptions().size();i++){
                        xy1.put(x.getKey().getOptions().get(i),0);
                    }
                    for (Response response: x.getValue()) {
                        try {
                            JSONArray res = new JSONArray(response.getResponse());
                            Log.e("multi ans", res.toString());
                            for(int i = 0; i <res.length(); i++){
                                if(xy1.containsKey(res.getString(i))){
                                    int j =  xy1.get(res.getString(i)) + 1;
                                    xy1.put(res.getString(i),j);
                                }
                                else{
                                    xy1.put(res.getString(i),1);
                                }
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    }
                    xyCoord.put(x.getKey(),xy1);
                    questions.add(x.getKey());
                    break;
            }

        }
    }

    private class CustomAdapter1 extends RecyclerView.Adapter<CustomAdapter1.ViewHolder> {
        List<Question> source;

        private CustomAdapter1(List<Question> source) {
            this.source = source;
        }

        private void add(Question responseHeader) {
            source.add(responseHeader);
        }

        @Override
        public CustomAdapter1.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new CustomAdapter1.ViewHolder(getActivity().getLayoutInflater().inflate(R.layout.bar_chart_view,parent, false));
        }

        @Override
        public void onBindViewHolder(CustomAdapter1.ViewHolder holder, int position) {
            holder.qu_no.setText( source.get(position).getQuestionNo()+"");
            holder.qu_text.setText(source.get(position).getQuestionText());
            switch(source.get(position).getQuestionType()){
                case "TEXT":
                    holder.barChart.setNoDataText("Bar Chart not available for this question type");
                    holder.btn.setVisibility(View.GONE);
                    break;
                default:
                    List<BarEntry> data = new ArrayList<>();
                    int i = 0;
                    boolean noData = true;
                    String[] labels = new String[xyCoord.get(source.get(position)).entrySet().size()];
                    List<Integer> colors = new ArrayList<>();
                    Log.e("sharp",xyCoord.get(source.get(position)).size()+"");
                    if(xyCoord.get(source.get(position)).size()<=0){
                        holder.barChart.setNoDataText("No response has been recorded for this question");
                        holder.btn.setVisibility(View.GONE);
                    }
                    else {
                        for (Map.Entry entry : xyCoord.get(source.get(position)).entrySet()) {
                            data.add(new BarEntry(i, ((int) entry.getValue())));
                            Random rnd = new Random();
                            int color = Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256));
                            colors.add(color);
                            labels[((int) (i))] = (String) entry.getKey();
                            i += 1;
                            if((int)entry.getValue()>0){
                                noData = false;
                            }
                        }
                        if(noData){
                            holder.barChart.setNoDataText("No response has been recorded for this question");
                            holder.btn.setVisibility(View.GONE);
                        }
                        else {
                            BarDataSet dataSet = new BarDataSet(data, "Responses");
                            dataSet.setColors(colors);
                            dataSet.setHighlightEnabled(true);
                            BarData barData = new BarData(dataSet);
                            barData.setValueTextSize(20);
                            barData.setBarWidth(0.5f);
                            barData.setValueFormatter(new IValueFormatter() {
                                @Override
                                public String getFormattedValue(float value, Entry entry, int dataSetIndex, ViewPortHandler viewPortHandler) {
                                    return entry.getY() == 0 ? "" : ((int) entry.getY()) + "";
                                }
                            });
                            holder.barChart.setData(barData);
                            holder.barChart.setPinchZoom(true);
                            holder.barChart.getAxisRight().setEnabled(false);
                            XAxis xaxis = holder.barChart.getXAxis();
                            YAxis yaxis = holder.barChart.getAxisLeft();
                            yaxis.setDrawGridLines(false);
                            xaxis.setPosition(XAxis.XAxisPosition.BOTTOM);
                            xaxis.setDrawGridLines(false);
                            xaxis.setDrawAxisLine(true);
                            xaxis.setGranularity(1f);
                            xaxis.setValueFormatter(new MyXAxisValueFormatter(labels));
                            //   holder.barChart.setMarker(new MyMarkerView(getContext(),R.layout.marker_view,labels));
                            holder.barChart.getDescription().setEnabled(false);
                            // Legend legend = holder.barChart.getLegend();
                            holder.barChart.animateXY(3000, 3000);
                        }
                    }
                    break;
            }


        }

        @Override
        public int getItemCount() {
            return source.size();
        }

        protected class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
            TextView qu_text;
            TextView qu_no;
            Button btn;
            BarChart barChart;

            private ViewHolder(View itemView) {
                super(itemView);
                qu_no = (TextView) itemView.findViewById(R.id.qu_no);
                qu_text = (TextView) itemView.findViewById(R.id.qu_text);
                barChart = (BarChart) itemView.findViewById(R.id.barChart);
                btn = (Button) itemView.findViewById(R.id.btn);
                btn.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        View view1 =  recyclerView.findViewHolderForAdapterPosition(getAdapterPosition()).itemView;
                         barChart1 = (BarChart)  view1.findViewById(R.id.barChart);
                        checkFilePermissionAndSave();
                    }
                });
                //itemView.setOnClickListener(this);
            }

            @Override
            public void onClick(View view) {
//                Intent intent = new Intent(getContext(), ResponseDetail.class);
//                intent.putExtra("survey_id",survey_id);
//                intent.putExtra("ID",source.get(getAdapterPosition()).getResponse_id());
//                startActivity(intent);
            }
        }
        protected class MyXAxisValueFormatter implements IAxisValueFormatter {

            private String[] mValues;

            public MyXAxisValueFormatter(String[] values) {
                this.mValues = values;
            }

            @Override
            public String getFormattedValue(float value, AxisBase axis) {
                // "value" represents the position of the label on the axis (x or y)
                Log.e("AXIS VALUE",value+"");
                return mValues[(int) value];
            }

            /** this is only needed if numbers are returned, else return 0 */

            public int getDecimalDigits() { return 0; }
        }
    }
    private class MyMarkerView extends MarkerView {

        private TextView tvContent;
        private String[] labels;

        public MyMarkerView(Context context, int layoutResource, String[] labels) {
            super(context, layoutResource);
            tvContent = (TextView) findViewById(R.id.marker);
         //   tvContent = (TextView) getActivity().getLayoutInflater().inflate(layoutResource,null).findViewById(R.id.marker);
            this.labels = labels;
            // find your layout components
        }

        // callbacks everytime the MarkerView is redrawn, can be used to update the
        // content (user-interface)
        @Override
        public void refreshContent(Entry e, Highlight highlight) {
            Log.e("refreshing marker: ",labels[(int)e.getX()-1]);
            tvContent.setText(labels[(int)e.getX()-1] + ": " + e.getY());

            // this will perform necessary layouting
            super.refreshContent(e, highlight);
        }

        private MPPointF mOffset;

        @Override
        public MPPointF getOffset() {

            if(mOffset == null) {
                // center the marker horizontally and vertically
                mOffset = new MPPointF(-(getWidth() / 2), -getHeight());
            }

            return mOffset;
        }
    }
    public boolean saveImage(BarChart barChart){
        return barChart.saveToGallery("bar_chart_"+new SimpleDateFormat("yyyy_MM_dd_HHmmss").format(new Date()),100);
    }
    public void checkFilePermissionAndSave(){
        if (ContextCompat.checkSelfPermission(getContext(),android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
                    android.Manifest.permission.WRITE_EXTERNAL_STORAGE)) {

            } else {

                // No explanation needed, we can request the permission.

                requestPermissions(
                        new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_PERMISSIONS_CODE_WRITE_STORAGE);

                // MY_PERMISSIONS_REQUEST_READ_CONTACTS is an
                // app-defined int constant. The callback method gets the
                // result of the request.
            }
        }
        else{
            if(barChart1 != null && saveImage(barChart1)){
                Toast.makeText(getContext(), "Chart saved", Toast.LENGTH_SHORT).show();
            }
            else {
                Toast.makeText(getContext(), "Chart failed to save", Toast.LENGTH_SHORT).show();
            }
        }
    }
    @Override
    public void onRequestPermissionsResult(int requestCode,  String[] permissions, int[] grantResults) {
        if (requestCode == REQUEST_PERMISSIONS_CODE_WRITE_STORAGE) {
            if (permissions[0].equals(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if(barChart1 != null && saveImage(barChart1)){
                    Toast.makeText(getContext(), "Chart saved", Toast.LENGTH_SHORT).show();
                }
                else {
                    Toast.makeText(getContext(), "Chart failed to save", Toast.LENGTH_SHORT).show();
                }
            }
            else{
                Toast.makeText(getContext(), "Chart failed to save", Toast.LENGTH_SHORT).show();
            }
        }

    }
}

File: app/src/main/java/com/example/babajidemustapha/survey/features/Reports/fragments/BarChartFragment.kt
package com.example.babajidemustapha.survey.features.Reports.fragments

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.babajidemustapha.survey.R
import com.example.babajidemustapha.survey.shared.models.QuestionType
import com.example.babajidemustapha.survey.shared.room.db.SurveyDatabase
import com.example.babajidemustapha.survey.shared.room.entities.Question
import com.example.babajidemustapha.survey.shared.room.entities.ResponseDetail
import com.example.babajidemustapha.survey.shared.utils.DbOperationHelper
import com.example.babajidemustapha.survey.shared.utils.DbOperationHelper.IDbOperationHelper
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.components.MarkerView
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.formatter.IAxisValueFormatter
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.utils.MPPointF
import org.json.JSONArray
import org.json.JSONException
import java.text.SimpleDateFormat
import java.util.*

/**
 * A simple [Fragment] subclass.
 */
class BarChartFragment : Fragment() {
    var survey_id = 0
    var db: SurveyDatabase? = null
    lateinit var recyclerView: RecyclerView
    var questions: MutableList<Question> = mutableListOf()
    var xyCoord: MutableMap<Question, Map<String, Int>?> = mutableMapOf()
    private val REQUEST_PERMISSIONS_CODE_WRITE_STORAGE = 18880
    var barChart1: BarChart? = null
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        barChart1 = null
        survey_id = activity!!.intent.extras.getInt("ID")
        db = SurveyDatabase.Companion.getInstance(context)
        questions = ArrayList()
        xyCoord = LinkedHashMap()
        recyclerView = view.findViewById(R.id.barList)
        recyclerView.layoutManager = LinearLayoutManager(context)
        DbOperationHelper.execute<Map<Question, List<ResponseDetail>>>(object : IDbOperationHelper<Map<Question, List<ResponseDetail>>> {
            override fun run(): Map<Question, List<ResponseDetail>> {
                return db!!.surveyDao().generateReport(survey_id)
            }

            override fun onCompleted(result: Map<Question, List<ResponseDetail>>) {
                buildXYcoord(result)
                recyclerView.adapter = CustomAdapter1(questions)
            }
        })
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_bar_chart, container, false)
    }

    fun buildXYcoord(report: Map<Question, List<ResponseDetail>>) {
        for ((key, value) in report) {

            when (key.questionType) {
                QuestionType.SHORT_TEXT -> {
                    xyCoord[key] = null
                    questions.add(key)
                }
                QuestionType.SINGLE_OPTION -> {
                    val xy: MutableMap<String, Int> = LinkedHashMap()
                    var i = 0
                    while (i < key.options!!.size) {
                        xy[key.options!![i]] = 0
                        i++
                    }
                    for (responseDetail in value) {
                        if (responseDetail.response != null && !responseDetail.response!!.trim { it <= ' ' }.isEmpty()) {
                            if (xy.containsKey(responseDetail.response!!)) {
                                val i = xy[responseDetail.response!!]!! + 1
                                xy[responseDetail.response!!] = i
                            } else {
                                xy[responseDetail.response!!] = 1
                            }
                        }
                    }
                    xyCoord[key] = xy
                    questions.add(key)
                }
                QuestionType.MULTIPLE_OPTION -> {
                    val xy1: MutableMap<String, Int> = LinkedHashMap()
                    var i = 0
                    while (i < key.options!!.size) {
                        xy1[key.options!![i]] = 0
                        i++
                    }
                    for (responseDetail in value) {
                        try {
                            val res = JSONArray(responseDetail.response)
                            Log.e("multi ans", res.toString())
                            var i = 0
                            while (i < res.length()) {
                                if (xy1.containsKey(res.getString(i))) {
                                    val j = xy1[res.getString(i)]!! + 1
                                    xy1[res.getString(i)] = j
                                } else {
                                    xy1[res.getString(i)] = 1
                                }
                                i++
                            }
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }
                    }
                    xyCoord[key] = xy1
                    questions.add(key)
                }
            }
        }
    }

    inner class CustomAdapter1(var source: MutableList<Question>) : RecyclerView.Adapter<CustomAdapter1.ViewHolder>() {
        private fun add(responseHeader: Question) {
            source.add(responseHeader)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(activity!!.layoutInflater.inflate(R.layout.bar_chart_view, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.qu_no.text = source[position].questionNo.toString()
            holder.qu_text.text = source[position].questionText
            when (source[position].questionType) {
                QuestionType.SHORT_TEXT -> {
                    holder.barChart.setNoDataText("Bar Chart not available for this question type")
                    holder.btn.visibility = View.GONE
                }
                else -> {
                    val data: MutableList<BarEntry> = ArrayList()
                    var i = 0
                    var noData = true
                    val labels = arrayOf<String>(xyCoord[source[position]]!!.entries.size.toString())
                    val colors: MutableList<Int> = ArrayList()
                    Log.e("sharp", xyCoord[source[position]]!!.size.toString() + "")
                    if (xyCoord[source[position]]!!.isEmpty()) {
                        holder.barChart.setNoDataText("No response has been recorded for this question")
                        holder.btn.visibility = View.GONE
                    } else {
                        for ((key, value) in xyCoord[source[position]]!!) {
                            data.add(BarEntry(i.toFloat(), value.toFloat()))
                            val rnd = Random()
                            val color = Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256))
                            colors.add(color)
                            labels[i] = key
                            i += 1
                            if (value > 0) {
                                noData = false
                            }
                        }
                        if (noData) {
                            holder.barChart.setNoDataText("No response has been recorded for this question")
                            holder.btn.visibility = View.GONE
                        } else {
                            val dataSet = BarDataSet(data, "Responses")
                            dataSet.colors = colors
                            dataSet.isHighlightEnabled = true
                            val barData = BarData(dataSet)
                            barData.setValueTextSize(20f)
                            barData.barWidth = 0.5f
                            barData.setValueFormatter { value, entry, dataSetIndex, viewPortHandler -> if (entry.y == 0f) "" else entry.y.toInt().toString() + "" }
                            holder.barChart.data = barData
                            holder.barChart.setPinchZoom(true)
                            holder.barChart.axisRight.isEnabled = false
                            val xaxis = holder.barChart.xAxis
                            val yaxis = holder.barChart.axisLeft
                            yaxis.setDrawGridLines(false)
                            xaxis.position = XAxis.XAxisPosition.BOTTOM
                            xaxis.setDrawGridLines(false)
                            xaxis.setDrawAxisLine(true)
                            xaxis.granularity = 1f
                            xaxis.valueFormatter = MyXAxisValueFormatter(labels)
                            //   holder.barChart.setMarker(new MyMarkerView(getContext(),R.layout.marker_view,labels));
                            holder.barChart.description.isEnabled = false
                            // Legend legend = holder.barChart.getLegend();
                            holder.barChart.animateXY(3000, 3000)
                        }
                    }
                }
            }
        }

        override fun getItemCount(): Int {
            return source.size
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
            var qu_text: TextView
            var qu_no: TextView
            var btn: Button
            var barChart: BarChart
            override fun onClick(view: View) {
//                Intent intent = new Intent(getContext(), ResponseDetail.class);
//                intent.putExtra("survey_id",survey_id);
//                intent.putExtra("ID",source.get(getAdapterPosition()).getResponse_id());
//                startActivity(intent);
            }

            init {
                qu_no = itemView.findViewById(R.id.qu_no)
                qu_text = itemView.findViewById(R.id.qu_text)
                barChart = itemView.findViewById(R.id.barChart)
                btn = itemView.findViewById(R.id.btn)
                btn.setOnClickListener {
                    val view1 = recyclerView.findViewHolderForAdapterPosition(adapterPosition)!!.itemView
                    barChart1 = view1.findViewById(R.id.barChart)
                    checkFilePermissionAndSave()
                }
                //itemView.setOnClickListener(this);
            }
        }

        protected inner class MyXAxisValueFormatter(private val mValues: Array<String>) : IAxisValueFormatter {
            override fun getFormattedValue(value: Float, axis: AxisBase): String {
                // "value" represents the position of the label on the axis (x or y)
                Log.e("AXIS VALUE", value.toString() + "")
                return mValues[value.toInt()]
            }

            /** this is only needed if numbers are returned, else return 0  */
            val decimalDigits: Int
                get() = 0
        }
    }

    private inner class MyMarkerView(context: Context?, layoutResource: Int, labels: Array<String>) : MarkerView(context, layoutResource) {
        private val tvContent: TextView
        private val labels: Array<String>

        // callbacks everytime the MarkerView is redrawn, can be used to update the
        // content (user-interface)
        override fun refreshContent(e: Entry, highlight: Highlight) {
            Log.e("refreshing marker: ", labels[e.x.toInt() - 1])
            tvContent.text = labels[e.x.toInt() - 1] + ": " + e.y

            // this will perform necessary layouting
            super.refreshContent(e, highlight)
        }

        private var mOffset: MPPointF? = null
        override fun getOffset(): MPPointF {
            if (mOffset == null) {
                // center the marker horizontally and vertically
                mOffset = MPPointF((-(width / 2)).toFloat(), (-height).toFloat())
            }
            return mOffset!!
        }

        init {
            tvContent = findViewById(R.id.marker)
            //   tvContent = (TextView) getActivity().getLayoutInflater().inflate(layoutResource,null).findViewById(R.id.marker);
            this.labels = labels
            // find your layout components
        }
    }

    fun saveImage(barChart: BarChart): Boolean {
        return barChart.saveToGallery("bar_chart_" + SimpleDateFormat("yyyy_MM_dd_HHmmss").format(Date()), 100)
    }

    fun checkFilePermissionAndSave() {
        if (ContextCompat.checkSelfPermission(context!!, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(activity!!,
                            Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            } else {

                // No explanation needed, we can request the permission.
                requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), REQUEST_PERMISSIONS_CODE_WRITE_STORAGE)

                // MY_PERMISSIONS_REQUEST_READ_CONTACTS is an
                // app-defined int constant. The callback method gets the
                // result of the request.
            }
        } else {
            if (barChart1 != null && saveImage(barChart1!!)) {
                Toast.makeText(context, "Chart saved", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(context, "Chart failed to save", Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == REQUEST_PERMISSIONS_CODE_WRITE_STORAGE) {
            if (permissions[0] == Manifest.permission.WRITE_EXTERNAL_STORAGE && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (barChart1 != null && saveImage(barChart1!!)) {
                    Toast.makeText(context, "Chart saved", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(context, "Chart failed to save", Toast.LENGTH_SHORT).show()
                }
            } else {
                Toast.makeText(context, "Chart failed to save", Toast.LENGTH_SHORT).show()
            }
        }
    }
}